### PR TITLE
Split things out so that we have a separate turtlebot2_demo package.

### PR DESCRIPTION
**Pull request by clalancette**
That way, turtlebot2_drivers doesn't depend on a lot of things
that don't belong in it.

The turtlebot2_demo package only has the launch files, and only does an install of those.  That means the turtlebot2_drivers package now just depends on the things it truly needs.

### CODE REVIEW

**Comment by clalancette** [comment]
This isn't quite ready; I realized there are a couple more dependencies I should add here.  I'll switch to "review" once it is truly ready.

**Comment by clalancette** [comment]
@nuclearsandwich I'm mostly looking for your review here to see if there are any obvious dependency issues.  We won't know for certain until you go to build, but keep in mind that we will now have a new package after this.

CI:
* TB2 Linux [Build Status](http://ci.ros2.org/job/ci_turtlebot-demo_linux/61/)
* TB2 Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_turtlebot-demo_linux-aarch64/77/)

**Comment by nuclearsandwich** [comment]
> I'm mostly looking for your review here to see if there are any obvious dependency issues.

Does `turtlebot2_demo` rely on any kinetic packages that aren't dependencies of `turtlebot2_drivers`?

**Comment by nuclearsandwich** [comment]
I think the big thing is that turtlebot2_demo will also need to source kinetic's workspace at build- and run-time in order to make use of turtlebot2_drivers. It'll be the first time that a bridged package (tentative way I'm referring to packages that require both ROS 1 and ROS 2) depends on another bridged package so if we're gonna see a :boom: I expect we'll see it here.

**Comment by clalancette** [comment]
No, turtlebot2_demo only depends on ROS2 packages.  It expects that those ROS2 packages pull in whatever they need; in the case of turtlebot2_drivers, that is obviously the kobuki stuff from ROS1.  The new package should be very straightforward to package, assuming I've gotten the dependencies correct.

**Review by nuclearsandwich — APPROVED** [review]
When packaging and running this demo. We'll need to source both the kinetic and ROS 2 workspaces. There is nowhere to indicate that in the source or package metadata at the moment, but we will need to document that for end users.

**Comment by clalancette** [comment]
@nuclearsandwich Thanks for the approval.  I'll wait a bit longer here to see if @mikaelarguedas or anyone else has comments about the name "turtlebot2_demo" or anything else, then I'll merge.

**Comment by mikaelarguedas** [comment]
(Haven't look at the change yetjust have a few high level questions first)
Is turtlebot2_demo supposed to be the package with all the launchfile of all demos? or only the cartographer one? Should amcl [launchfiles](https://github.com/ros2/turtlebot2_demo/pull/44) land in that new packages or somewhere else ?

If it's only for the cartographer files it should be renamed, if it's for all of them we should wait for https://github.com/ros2/turtlebot2_demo/pull/44 and refactor that along.

**Comment by mikaelarguedas** [comment]
(I commented in a hurry because I was concerned that this would get merged right away, I'll look at the code and update my comment accordingly)

**Comment by clalancette** [comment]
> (Haven't look at the change yetjust have a few high level questions first)
> Is turtlebot2_demo supposed to be the package with all the launchfile of all demos? or only the cartographer one? Should amcl launchfiles land in that new packages or somewhere else ?

It's for all of them; I included the follower and joy launch files as well.  I don't particularly care which order we merge this and #44, though there is clearly going to be a conflict there.

**Comment by dhood** [comment]
I would think they'd be separate packages so you don't pull in cartographer dependencies unless you need to. In my mind the drivers package would include a 'common' launch file + dependencies, then cartographer demo just needs to depend on drivers + cartographer_ros

**Review by wjwwood — APPROVED** [review]
lgtm too, but I think the comment @mikaelarguedas made about the amcl launch files should be resolved too.

**Comment by clalancette** [comment]
> I would think they'd be separate packages so you don't pull in cartographer dependencies unless you need to. In my mind the drivers package would include a 'common' launch file + dependencies, then cartographer demo just needs to depend on drivers + cartographer_ros

True, but I think having too many packages is also confusing.  I think having a sort of "meta-package" where it pulls in all of the dependencies for all of the demos isn't a bad thing to have.

**Comment by mikaelarguedas** [comment]
If it's a package will all the demos we just group all the dependencies into one massive package. While preventing `turtlebot2_drivers` to depend on very non driver related packages, I think it's not what the only goal of this refactor. That's my bad I should have bullet pointed what I was planning on doing here and get feedback before getting you started on it :S.

I was thinking of a bunch of demo specific packages with a "meta-package" on top for convenience. (e.g have one `turtlebot2_cartographer`, one `turtlebot2_amcl`, one `tutlebot2_teleop` and a `turtlebot_follower` and finally a metapackage `turtlebot2_demos` depending on all of them).

Though if the team think it's good with the current structure and can be improved later on. I think that it should be called turtlebot2_demo**s** and at least declare all the dependencies (didnt look closely but at least `ament_index_python` is missing)

Then part of this refactor was also to follow_up of https://github.com/ros2/ros2/issues/364 and install every executable where `ros2 run` is looking for them and update the launchfiles to use that.

**Comment by clalancette** [comment]
So after this last comment, and discussion on slack, I'm going to change this to use the broken-out packages.  I'm less certain about the `ros2 run` portion of this; once I have the sub-packages, I'll ask around about that bit if we still have time.

**Comment by mikaelarguedas** [comment]
I think that we need to be consistent, the entire stack has been converted so that every executable is installed in the expected location. Can you explain the rationale for having only the turtlebot2_demo executables on your path after installing ros2 ?

**Comment by clalancette** [comment]
The rationale is that I don't know how to do it, and we are running out of time :).  I'll try to do it once I'm done with the packages.

File: turtlebot2_demo/launch/turtlebot_joy.py [hunk 1]
```diff
@@ -23,8 +23,12 @@ def launch(launch_descriptor, argv):
...
+        cmd=['teleop_node'],
+        name='teleop_node',
+        exit_handler=restart_exit_handler,
+    )
+    ld.add_process(
         cmd=['joy_node'],
         name='joy_node',
-        # The joy node is required, die if it dies
-        exit_handler=default_exit_handler,
+        exit_handler=restart_exit_handler,
```
Comment (dhood) [inline turtlebot2_demo/launch/turtlebot_joy.py]: I ran into a situation where use of restart_exit_handler in another launch file caused the turtlebot to lock up. I passed an invalid argument to map_server and it kept respawning and dying repeatedly. Perhaps we should be sparing with the use of the restart exit handler..!

**Comment by mikaelarguedas** [comment]
Fine I'll do it

**Comment by clalancette** [comment]
I still need to test this new change, but this is the direction I'm going.  Scream if you don't like something in it.

**Comment by mikaelarguedas** [comment]
Don't merge it I'll iterate on it tonight

**Comment by clalancette** [comment]
All right.  So after all of our discussions, I was finally able to get launch files running from ros2 run by adding #! at the beginning, moving the launch files (and all of the binaries) to lib/${PROJECT_NAME}, making the launch files executable, and adding some code at the bottom of each launch file to actually launch.  Now, if you checkout the 'tb2-demo' branch on all of the repositories, you should be able to run:

```
$ ros2 run tb2_demo_carto turtlebot_carto_2d.py
```

And have it actually launch the demo.  There are a few problems, though:
1.  I have not moved tf2_ros static_transform_publisher into libexec yet.
1.  When running from ros2 run, the end of the output is truncated when hitting Ctrl-C.  That is, you never see the last stuff that the nodes printed.  This wasn't the case when using 'launch' previously.
1.  Hitting Ctrl-C doesn't always kill all of the nodes.  It sometimes kills them all, and sometimes leaves one (random) one running.

Those last two problems make it much less pleasant to work with than 'launch', but it might be good enough for now.  I'm going to give up on this for now; @mikaelarguedas , @wjwwood , feel free to take a look at all of the tb2-demo branches.  If you need me, please feel free to text me; I may not be around the computer too much tonight.

**Comment by mikaelarguedas** [comment]
thanks @clalancette for pounding on this! I'll try to have a look soonish

**Comment by mikaelarguedas** [comment]
I pushed my changes on [this branch](https://github.com/ros2/turtlebot2_demo/tree/tb2_repos_refactor)
What's in there:
 - removing dep on teleop_twist_keyboard because no launchfile uses it
 - rename the packages to have a consistent naming scheme, also renamed the tb2_demo_carto to use the full name cartographer so that user dont confuse with the [karto](http://wiki.ros.org/karto) planning library pretty popular in ROS1
- use the new path of `static_transform_publisher`
- merge turtlebot2_follower and tb2_demo_follower into a single package (taking example of the turtlebot teleop package)
- making teleop node optional for the follower demo (given that joy was already optional)

What's still missing: 
- I must be missing an ament_index_python dependency in at least the follower package maybe more
- Actual testing on the turtlebot to make sure all goes well
- Try to replace all nodes to not restart by default and see if nodes do tear down properly. Right now we are not sure if nodes cleanup properly after themselves when crashing and restating them automatically may lead to funky behavior at the dds level.

I'll address all this tomorrow morning

**Comment by mikaelarguedas** [comment]
These commits are now part of #46

**Pull request closed without merging**